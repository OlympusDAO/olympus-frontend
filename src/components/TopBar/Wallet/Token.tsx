import { t } from "@lingui/macro";
import {
  Accordion as MuiAccordion,
  AccordionDetails,
  AccordionSummary as MuiAccordionSummary,
  Box,
  Button,
  Typography,
  useTheme,
  withStyles,
} from "@material-ui/core";
import { Skeleton } from "@material-ui/lab";
import { Icon, OHMTokenProps, Token as TokenSVG } from "@olympusdao/component-library";
import { ChangeEvent, useState } from "react";
import { useQuery } from "react-query";
import { addresses, NETWORKS } from "src/constants";
import { NetworkId } from "src/constants";
import { formatCurrency } from "src/helpers";
import { useAppSelector } from "src/hooks";
import { useWeb3Context } from "src/hooks/web3Context";
import { fetchCrossChainBalances } from "src/lib/fetchBalances";

import { trackGAEvent, trackSegmentEvent } from "../../../helpers/analytics";

const Accordion = withStyles({
  root: {
    backgroundColor: "inherit",
    backdropFilter: "none",
    "-webkit-backdrop-filter": "none",
    boxShadow: "none",
    "&:before": {
      display: "none",
    },
    "&$expanded": {
      margin: 0,
    },
  },
  expanded: {},
})(MuiAccordion);

const AccordionSummary = withStyles(theme => ({
  root: {
    minHeight: "36px",
    height: "36px",
    padding: theme.spacing(0),
    "&$expanded": {
      padding: theme.spacing(0),
      minHeight: "36px",
    },
  },
  content: {
    margin: 0,
    "&$expanded": {
      margin: 0,
    },
  },
  expanded: {},
}))(MuiAccordionSummary);

export interface IToken {
  symbol: string;
  address: string;
  decimals: number;
  icon: OHMTokenProps["name"];
  balance: string;
  price: number;
  crossChainBalances?: { balances: Record<NetworkId, string>; isLoading: boolean };
  vaultBalances?: { [vaultName: string]: string };
  totalBalance: string;
}

const addTokenToWallet = async (token: IToken, userAddress: string) => {
  if (!window.ethereum) return;
  const host = window.location.origin;
  try {
    await window.ethereum.request({
      method: "wallet_watchAsset",
      params: {
        type: "ERC20",
        options: {
          address: token.address,
          symbol: token.symbol,
          decimals: token.decimals,
          image: `${host}/${token.icon}`,
        },
      },
    });
    trackSegmentEvent({
      address: userAddress,
      type: "Add Token",
      tokenName: token.symbol,
    });
    trackGAEvent({
      category: "Token",
      action: "Add Token",
      label: token.symbol ?? "unknown",
    });
  } catch (error) {
    console.log(error);
  }
};

interface TokenProps extends IToken {
  expanded: boolean;
  onChangeExpanded: (event: ChangeEvent<any>, isExpanded: boolean) => void;
  onAddTokenToWallet: () => void;
  decimals: number;
}

const BalanceValue = ({
  balance,
  balanceValueUSD,
  isLoading = false,
  sigFigs,
}: {
  balance: string;
  balanceValueUSD: number;
  isLoading?: boolean;
  sigFigs: number;
}) => (
  <Box sx={{ textAlign: "right", display: "flex", flexDirection: "column", justifyContent: "space-between" }}>
    <Typography variant="body2" style={{ fontWeight: 600 }}>
      {!isLoading ? balance.substring(0, sigFigs) : <Skeleton variant="text" width={50} />}
    </Typography>
    <Typography variant="body2" color="textSecondary">
      {!isLoading ? (
        formatCurrency(balanceValueUSD === NaN ? 0 : balanceValueUSD, 2)
      ) : (
        <Skeleton variant="text" width={50} />
      )}
    </Typography>
  </Box>
);

const TokenBalance = ({
  balanceLabel,
  balance,
  balanceValueUSD,
  sigFigs,
}: {
  balanceLabel: string;
  balance: string;
  balanceValueUSD: number;
  sigFigs: number;
}) => (
  <Box display="flex" flexDirection="row" justifyContent="space-between" key={balanceLabel}>
    <Typography color="textSecondary">{balanceLabel}</Typography>
    <Typography color="textSecondary">
      <BalanceValue balance={balance} sigFigs={sigFigs} balanceValueUSD={balanceValueUSD} />
    </Typography>
  </Box>
);

export const Token = ({
  symbol,
  decimals,
  icon,
  price = 0,
  crossChainBalances,
  vaultBalances,
  totalBalance,
  onAddTokenToWallet,
  expanded,
  onChangeExpanded,
}: TokenProps) => {
  const theme = useTheme();
  const isLoading = useAppSelector(s => s.account.loading || s.app.loadingMarketPrice || s.app.loading);
  const balanceValue = parseFloat(totalBalance) * price;

  // cleanedDecimals provides up to 7 sigFigs on an 18 decimal token (gOHM) & 5 sigFigs on 9 decimal Token
  const sigFigs = decimals === 18 ? 7 : 5;

  return (
    <Accordion expanded={expanded} onChange={onChangeExpanded}>
      <AccordionSummary expandIcon={<Icon name="more" color="disabled" />}>
        <Box sx={{ display: "flex", alignItems: "center", width: "100%" }}>
          <TokenSVG name={icon} style={{ fontSize: 28, marginRight: theme.spacing(1) }} />
          <Typography>{symbol}</Typography>
        </Box>
        <BalanceValue
          balance={totalBalance}
          sigFigs={sigFigs}
          balanceValueUSD={balanceValue}
          isLoading={isLoading || crossChainBalances?.isLoading}
        />
      </AccordionSummary>
      <AccordionDetails style={{ margin: "auto", padding: theme.spacing(1, 0) }}>
        <Box
          sx={{ display: "flex", flexDirection: "column", flex: 1, mx: "32px", justifyContent: "center" }}
          style={{ gap: theme.spacing(1) }}
        >
          {!!crossChainBalances?.balances &&
            Object.entries(crossChainBalances.balances).map(
              ([networkId, balance]) =>
                parseFloat(balance) > 0.01 && (
                  <TokenBalance
                    balanceLabel={`${NETWORKS[networkId as any].chainName}:`}
                    balance={balance}
                    balanceValueUSD={parseFloat(balance) * price}
                    sigFigs={sigFigs}
                  />
                ),
            )}
          {!!vaultBalances &&
            Object.entries(vaultBalances).map(
              ([vaultName, balance]) =>
                parseFloat(balance) > 0.01 && (
                  <TokenBalance
                    balanceLabel={`${vaultName}:`}
                    balance={balance}
                    balanceValueUSD={parseFloat(balance) * price}
                    sigFigs={sigFigs}
                  />
                ),
            )}
          <Box className="ohm-pairs" style={{ width: "100%" }}>
            <Button variant="contained" color="secondary" fullWidth onClick={onAddTokenToWallet}>
              <Typography>{t`Add to Wallet`}</Typography>
            </Button>
          </Box>
        </Box>
      </AccordionDetails>
    </Accordion>
  );
};

export const MigrateToken = ({ symbol, icon, balance = "0.0", price = 0 }: IToken) => {
  const theme = useTheme();
  const balanceValue = parseFloat(balance) * price;
  return (
    <Box sx={{ display: "flex", flexDirection: "row" }}>
      <Box sx={{ display: "flex", alignItems: "center", width: "100%" }}>
        <TokenSVG name={icon} style={{ fontSize: 28, marginRight: theme.spacing(1) }} />
        <Typography>{symbol}</Typography>
      </Box>
      {/* <Button variant="contained" color="primary" size="small" onClick={() => true}>
        Migrate v2
      </Button> */}
      <Box
        sx={{
          mx: theme.spacing(0.5),
          textAlign: "right",
          display: "flex",
          flexDirection: "column",
          justifyContent: "space-between",
        }}
      >
        <Typography variant="body2" style={{ fontWeight: 600 }}>
          {balance.substring(0, 5)}
        </Typography>
        <Typography variant="body2" color="textSecondary">
          {formatCurrency(balanceValue, 2)}
        </Typography>
      </Box>
    </Box>
  );
};

const sumObjValues = (obj: Record<string, string> = {}) =>
  Object.values(obj).reduce((sum, b = "0.0") => sum + (parseFloat(b) || 0), 0);

export const useWallet = (
  userAddress: string,
  chainId: NetworkId,
  providerInitialized: boolean,
): Record<string, IToken> => {
  // default to mainnet while not initialized
  const networkId = providerInitialized ? chainId : NetworkId.MAINNET;

  const connectedChainBalances = useAppSelector(s => s.account.balances);
  const ohmPrice = useAppSelector(s => s.app.marketPrice);
  const currentIndex = useAppSelector(s => s.app.currentIndex);

  const { gohm, wsohm, isLoading } = useCrossChainBalances(userAddress);

  const tokens = {
    ohmV1: {
      symbol: "OHM V1",
      address: addresses[networkId].OHM_ADDRESS,
      balance: connectedChainBalances.ohmV1,
      price: ohmPrice || 0,
      icon: "OHM",
      decimals: 9,
    },
    sohmV1: {
      symbol: "sOHM V1",
      address: addresses[networkId].SOHM_ADDRESS,
      balance: connectedChainBalances.sohmV1,
      price: ohmPrice || 0,
      icon: "sOHM",
      decimals: 9,
    },
    ohm: {
      symbol: "OHM",
      address: addresses[networkId].OHM_V2,
      balance: connectedChainBalances.ohm,
      price: ohmPrice || 0,
      icon: "OHM",
      decimals: 9,
    },
    sohm: {
      symbol: "sOHM",
      address: addresses[networkId].SOHM_V2,
      balance: connectedChainBalances.sohm,
      price: ohmPrice || 0,
      vaultBalances: {
        "Fuse Olympus Pool Party": connectedChainBalances.fsohm,
      },
      icon: "sOHM",
      decimals: 9,
    },
    wsohm: {
      symbol: "wsOHM",
      address: addresses[networkId].WSOHM_ADDRESS,
      balance: connectedChainBalances.wsohm,
      price: (ohmPrice || 0) * Number(currentIndex || 0),
      crossChainBalances: { balances: wsohm, isLoading },
      icon: "wsOHM",
      decimals: 18,
    },
    pool: {
      symbol: "33T",
      address: addresses[networkId].PT_TOKEN_ADDRESS,
      balance: connectedChainBalances.pool,
      price: ohmPrice || 0,
      icon: "33T",
      decimals: 9,
    },
    gohm: {
      symbol: "gOHM",
      address: addresses[networkId].GOHM_ADDRESS,
      balance: connectedChainBalances.gohm,
      price: (ohmPrice || 0) * Number(currentIndex || 0),
      crossChainBalances: { balances: gohm, isLoading },
      vaultBalances: {
        "gOHM on Tokemak": connectedChainBalances.gOhmOnTokemak,
        "Fuse Olympus Pool Party": connectedChainBalances.fgohm,
      },
      icon: "wsOHM",
      decimals: 18,
    },
  } as Record<string, Omit<IToken, "totalBalance">>;

  return Object.entries(tokens).reduce((wallet, [key, token]) => {
    const crossChainBalances = sumObjValues(token.crossChainBalances?.balances);
    const vaultBalances = sumObjValues(token.vaultBalances);
    const balance = crossChainBalances || parseFloat(token.balance) || 0;
    return {
      ...wallet,
      [key]: {
        ...token,
        totalBalance: (balance + vaultBalances).toString(),
      } as IToken,
    };
  }, {});
};

export const useCrossChainBalances = (address: string) => {
  const { isLoading, data } = useQuery(["crossChainBalances", address], () => fetchCrossChainBalances(address), {
    refetchOnWindowFocus: false,
    refetchOnReconnect: false,
  });
  return { isLoading, ...data };
};

export const Tokens = () => {
  const { address: userAddress, networkId, providerInitialized } = useWeb3Context();
  const tokens = useWallet(userAddress, networkId, providerInitialized);
  const isLoading = useAppSelector(s => s.account.loading || s.app.loadingMarketPrice || s.app.loading);
  const [expanded, setExpanded] = useState<string | null>(null);

  const v1Tokens = [tokens.ohmV1, tokens.sohmV1];
  const alwaysShowTokens = [tokens.ohm, tokens.sohm, tokens.gohm];
  const onlyShowWhenBalanceTokens = [tokens.wsohm, tokens.pool];

  const tokenProps = (token: IToken) => ({
    ...token,
    expanded: expanded === token.symbol,
    onChangeExpanded: (e: any, isExpanded: boolean) => setExpanded(isExpanded ? token.symbol : null),
    onAddTokenToWallet: () => addTokenToWallet(token, userAddress),
  });

  return (
    <>
      {alwaysShowTokens.map(token => (
        <Token key={token.symbol} {...tokenProps(token)} />
      ))}
      {!isLoading &&
        onlyShowWhenBalanceTokens.map(
          token => parseFloat(token.totalBalance) > 0.01 && <Token key={token.symbol} {...tokenProps(token)} />,
        )}
      {!isLoading &&
        v1Tokens.map(token => parseFloat(token.totalBalance) > 0.01 && <MigrateToken {...token} key={token.symbol} />)}
    </>
  );
};
