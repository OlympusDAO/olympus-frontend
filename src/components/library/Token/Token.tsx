import { SvgIcon, SvgIconProps } from "@mui/material";
import { styled } from "@mui/material/styles";
import { FC } from "react";
import tokenPath from "src/components/library/Token/tokensLib";

const PREFIX = "Token";

const classes = {
  root: `${PREFIX}-root`,
};

const StyledSvgIcon = styled(SvgIcon)(() => ({
  [`&.${classes.root}`]: {
    display: "flex",
    justifyContent: "space-between",
    flexDirection: "row",
    margin: "12px 0px",
  },
}));

export interface OHMTokenProps extends SvgIconProps {
  name: keyof typeof tokenPath;
  viewBox?: string;
  fontSize?: SvgIconProps["fontSize"];
}
/**
 * Primary Token Component for UI. This component displays a SVG Token Asset from the OlympusDAO Token Library.
 *
 * ## Updating Component Library with a New Token Asset
 * You may find that a specific token asset is not available inside of the component library package. To add a new token to component-library follow the following steps.
 * 
 * #### Step 1
 * Ensure your SVG file is optimized using svgo. Run ```npx svgo <yourasset.svg> ```
 * 
 * #### Step 2
 * We've tried to make sure default viewports, viewboxes and overall heights/widths are aligned to a standard 32x32 as a base default. 
 *  https://boxy-svg.com/ is a great resource to verify that the default width/height of the SVG is set properly and that the viewbox is set to 32x32. Once verified/modified, export the SVG and move to the next step.
 * 
 * #### Step 3
 * Convert your SVG to a react component by running ```npx @svgr/cli <YourProperlySized.svg>```. SVGR will automatically normalize your SVG and rename attributes to play nice with react. 
 * You'll see a react component output to the command line. Copy the output BETWEEN the starting and closing ```<svg></svg>``` tags. Exclude the starting/ending ```<svg>``` tags. 
 *
 *  #### Step 4
 * Add a new Key:Value to tokensLib.tsx with Key being ALL CAPS for a Token Symbol or all lowercase for a network/other type of token. 
 * Example
 * 
 * ```jsx
 *   AVAX: (
 *   <>
 *     <path
 *       style={{
 *         fill: "#fff",
 *         strokeWidth: 0.02130493,
 *       }}
 *       d="M6.104 5.475h19.771v17.981H6.104z"
 *     />
 *     <path
 *       d="M32 16c0 8.837-7.163 16-16 16S0 24.837 0 16 7.163 0 16 0s16 7.163 16 16Zm-20.534 6.367H8.361c-.653 0-.975 0-1.171-.126a.79.79 0 0 1-.358-.617c-.012-.232.15-.515.472-1.08L14.97 7.028c.326-.574.49-.86.7-.967a.791.791 0 0 1 .715 0c.208.106.373.393.7.967L18.66 9.78l.008.014c.353.615.531.927.61 1.255.086.358.086.735 0 1.093-.08.33-.256.644-.614 1.27l-4.027 7.119-.01.018c-.355.62-.535.935-.784 1.172-.271.26-.597.448-.955.555-.326.09-.692.09-1.423.09zm7.842 0h4.449c.656 0 .987 0 1.183-.13a.787.787 0 0 0 .358-.62c.011-.225-.146-.497-.455-1.03l-.033-.055-2.228-3.813-.026-.043c-.313-.53-.471-.797-.674-.9a.783.783 0 0 0-.711 0c-.205.106-.37.385-.696.947l-2.22 3.813-.009.013c-.325.56-.487.841-.476 1.071a.796.796 0 0 0 .358.622c.193.125.523.125 1.18.125z"
 *       style={{
 *         clipRule: "evenodd",
 *         fill: "#e84142",
 *         fillRule: "evenodd",
 *         strokeWidth: 0.02130493,
 *       }}
 *     />
 *   </>
 * ),
  ```
 *  
 * #### Step 5
 * Submit a PR to the component-library project, to be published in the next npm package release. 
 */

const Token: FC<OHMTokenProps> = ({ name, viewBox = "0 0 32 32", fontSize = "large", ...props }) => {
  return (
    <StyledSvgIcon viewBox={viewBox} fontSize={fontSize} {...props}>
      {tokenPath[name]}
    </StyledSvgIcon>
  );
};

export default Token;
export type TokenName = keyof typeof tokenPath;
