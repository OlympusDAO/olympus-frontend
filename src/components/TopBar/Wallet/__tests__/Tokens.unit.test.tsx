import { screen } from "@testing-library/react";

import { render } from "../../../../testUtils";
import { Tokens } from "../Token";

describe("<Tokens/>", () => {
  let container: HTMLElement;

  beforeEach(() => {
    const component = render(<Tokens />);
    container = component.container;
  });

  it("should render component", () => {
    expect(container).toMatchSnapshot();
  });

  it("should show default tokens", async () => {
    const gOhmElement = await screen.findByText(/gOhm/i);
    expect(gOhmElement).toBeInTheDocument();

    const sOhmElement = await screen.findByText(/sOhm/i);
    expect(sOhmElement).toBeInTheDocument();
  });
});
