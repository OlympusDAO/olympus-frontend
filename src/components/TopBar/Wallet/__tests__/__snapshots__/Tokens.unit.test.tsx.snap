// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Tokens/> should render component 1`] = `
<div>
  <div
    class="MuiPaper-root MuiAccordion-root WithStyles(ForwardRef(Accordion))-root-1 MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root-3"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content-4"
      >
        <div
          class="MuiBox-root MuiBox-root-6"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            style="font-size: 28px; margin-right: 8px;"
            viewBox="0 0 32 32"
          >
            <path
              d="M0 16C0 7.163 7.163 0 16 0s16 7.163 16 16-7.163 16-16 16S0 24.837 0 16Z"
              fill="#708B96"
            />
            <path
              clip-rule="evenodd"
              d="M17.536 23.04v.032h5.504v-1.984h-2.718c1.635-1.239 2.686-3.165 2.686-5.328 0-3.738-3.138-6.768-7.008-6.768s-7.008 3.03-7.008 6.768c0 2.163 1.05 4.089 2.686 5.328H8.96v1.984h5.504v-2.667c-2.005-.63-3.456-2.465-3.456-4.629 0-2.686 2.235-4.864 4.992-4.864s4.992 2.178 4.992 4.864c0 2.164-1.45 3.998-3.456 4.63v2.634Z"
              fill="#fff"
              fill-rule="evenodd"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            OHM
          </p>
        </div>
        <div
          class="MuiBox-root MuiBox-root-8"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
            data-testid="balance-token"
            style="font-weight: 600;"
          >
            0
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            data-testid="balance-usd"
          >
            $0.00
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M10.4 6.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm0 10.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm-1.75-7a1.75 1.75 0 1 0 3.5 0 1.75 1.75 0 0 0-3.5 0Z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
              style="margin: auto; padding: 8px 0px;"
            >
              <div
                class="MuiBox-root MuiBox-root-10"
                style="gap: 8px;"
              >
                <div
                  class="MuiBox-root MuiBox-root-11 ohm-pairs"
                  style="width: 100%;"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-disableElevation MuiButton-fullWidth"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        Add to Wallet
                      </p>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiAccordion-root WithStyles(ForwardRef(Accordion))-root-1 MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root-3"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content-4"
      >
        <div
          class="MuiBox-root MuiBox-root-12"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            style="font-size: 28px; margin-right: 8px;"
            viewBox="0 0 32 32"
          >
            <defs>
              <lineargradient
                gradientTransform="matrix(0.319996, 0, 0, 0.319996, -0.000007, -1.775952)"
                gradientUnits="userSpaceOnUse"
                id="paint0_linear"
                x1="-83.5339"
                x2="189.059"
                y1="50"
                y2="50"
              >
                <stop
                  stop-color="#708B96"
                />
                <stop
                  offset="1"
                  stop-color="#F7FBE7"
                />
              </lineargradient>
            </defs>
            <path
              clip-rule="evenodd"
              d="M 15.982 28.228 C 23.688 28.228 29.934 21.982 29.934 14.276 C 29.934 12.658 29.658 11.104 29.152 9.659 L 29.152 5.11 C 30.947 7.696 32 10.837 32 14.224 C 32 23.06 24.836 30.224 16 30.224 C 7.163 30.224 0 23.06 0 14.224 C 0 10.817 1.065 7.659 2.88 5.064 L 2.88 9.472 C 2.331 10.97 2.031 12.588 2.031 14.276 C 2.031 21.982 8.276 28.228 15.982 28.228 Z"
              fill="url(#paint0_linear)"
              fill-rule="evenodd"
            />
            <path
              d="M 3.552 14.224 C 3.552 7.349 9.125 1.776 16 1.776 C 22.875 1.776 28.448 7.349 28.448 14.224 C 28.448 21.099 22.875 26.671 16 26.671 C 9.125 26.671 3.552 21.099 3.552 14.224 Z"
              fill="#708B96"
            />
            <path
              clip-rule="evenodd"
              d="M 17.195 19.176 C 17.195 19.176 17.195 19.176 17.195 19.176 L 17.195 19.726 L 21.477 19.726 L 21.477 18.181 L 19.363 18.181 C 20.634 17.219 21.451 15.72 21.451 14.037 C 21.451 11.129 19.011 8.772 16 8.772 C 12.989 8.772 10.548 11.129 10.548 14.037 C 10.548 15.72 11.365 17.219 12.637 18.181 L 10.523 18.181 L 10.523 19.726 L 14.805 19.726 L 14.805 19.176 L 14.805 17.652 C 13.245 17.16 12.116 15.733 12.116 14.05 C 12.116 11.96 13.855 10.266 16 10.266 C 18.145 10.266 19.884 11.96 19.884 14.05 C 19.884 15.733 18.755 17.16 17.195 17.652 L 17.195 19.176 Z"
              fill="white"
              fill-rule="evenodd"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            sOHM
          </p>
        </div>
        <div
          class="MuiBox-root MuiBox-root-13"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
            data-testid="balance-token"
            style="font-weight: 600;"
          >
            0
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            data-testid="balance-usd"
          >
            $0.00
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M10.4 6.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm0 10.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm-1.75-7a1.75 1.75 0 1 0 3.5 0 1.75 1.75 0 0 0-3.5 0Z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
              style="margin: auto; padding: 8px 0px;"
            >
              <div
                class="MuiBox-root MuiBox-root-14"
                style="gap: 8px;"
              >
                <div
                  class="MuiBox-root MuiBox-root-15 ohm-pairs"
                  style="width: 100%;"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-disableElevation MuiButton-fullWidth"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        Add to Wallet
                      </p>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="MuiPaper-root MuiAccordion-root WithStyles(ForwardRef(Accordion))-root-1 MuiAccordion-rounded MuiPaper-elevation0 MuiPaper-rounded"
  >
    <div
      aria-disabled="false"
      aria-expanded="false"
      class="MuiButtonBase-root MuiAccordionSummary-root WithStyles(ForwardRef(AccordionSummary))-root-3"
      role="button"
      tabindex="0"
    >
      <div
        class="MuiAccordionSummary-content WithStyles(ForwardRef(AccordionSummary))-content-4"
      >
        <div
          class="MuiBox-root MuiBox-root-16"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge"
            focusable="false"
            style="font-size: 28px; margin-right: 8px;"
            viewBox="0 0 32 32"
          >
            <defs>
              <lineargradient
                gradientTransform="matrix(0.177778, 0, 0, 0.177778, 1.777779, 1.777779)"
                gradientUnits="userSpaceOnUse"
                id="paint0_linear_359"
                x1="80"
                x2="80"
                y1="-84"
                y2="256"
              >
                <stop
                  offset="0.1949"
                  stop-color="#708B96"
                />
                <stop
                  offset="1"
                  stop-color="#F7FBE7"
                />
              </lineargradient>
            </defs>
            <path
              d="M 1.778 16 C 1.778 8.145 8.145 1.778 16 1.778 C 23.855 1.778 30.222 8.145 30.222 16 C 30.222 23.855 23.855 30.222 16 30.222 C 8.145 30.222 1.778 23.855 1.778 16 Z"
              fill="#fff"
            />
            <rect
              fill="#fff"
              height="12.516"
              width="12.516"
              x="9.742"
              y="9.771"
            />
            <path
              clip-rule="evenodd"
              d="M 14.635 22.286 L 14.635 19.916 C 12.852 19.355 11.563 17.725 11.563 15.801 C 11.563 13.413 13.549 11.477 16 11.477 C 18.451 11.477 20.437 13.413 20.437 15.801 C 20.437 17.725 19.148 19.355 17.365 19.916 L 17.365 21.658 L 17.365 22.258 L 17.365 22.286 L 22.258 22.286 L 22.258 20.523 L 19.842 20.523 C 21.295 19.421 22.229 17.709 22.229 15.787 C 22.229 12.464 19.44 9.771 16 9.771 C 12.56 9.771 9.771 12.464 9.771 15.787 C 9.771 17.709 10.705 19.421 12.158 20.523 L 9.742 20.523 L 9.742 22.286 Z"
              fill="#708b96"
              fill-rule="evenodd"
            />
            <path
              d="M 16 28.444 C 9.127 28.444 3.556 22.873 3.556 16 L 0 16 C 0 24.837 7.163 32 16 32 Z M 28.444 16 C 28.444 22.873 22.873 28.444 16 28.444 L 16 32 C 24.837 32 32 24.837 32 16 Z M 16 3.556 C 22.873 3.556 28.444 9.127 28.444 16 L 32 16 C 32 7.163 24.837 0 16 0 Z M 16 0 C 7.163 0 0 7.163 0 16 L 3.556 16 C 3.556 9.127 9.127 3.556 16 3.556 Z"
              fill="url(#paint0_linear_359)"
            />
          </svg>
          <p
            class="MuiTypography-root MuiTypography-body1"
          >
            gOHM
          </p>
        </div>
        <div
          class="MuiBox-root MuiBox-root-17"
        >
          <p
            class="MuiTypography-root MuiTypography-body2"
            data-testid="balance-token"
            style="font-weight: 600;"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
              style="width: 50px;"
            />
          </p>
          <p
            class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
            data-testid="balance-usd"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse"
              style="width: 50px;"
            />
          </p>
        </div>
      </div>
      <div
        aria-disabled="false"
        aria-hidden="true"
        class="MuiButtonBase-root MuiIconButton-root MuiAccordionSummary-expandIcon MuiIconButton-edgeEnd"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-colorDisabled MuiSvgIcon-fontSizeSmall"
            focusable="false"
            viewBox="0 0 20 20"
          >
            <path
              d="M10.4 6.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm0 10.5a1.75 1.75 0 1 1 0-3.5 1.75 1.75 0 0 1 0 3.5Zm-1.75-7a1.75 1.75 0 1 0 3.5 0 1.75 1.75 0 0 0-3.5 0Z"
            />
          </svg>
        </span>
      </div>
    </div>
    <div
      class="MuiCollapse-root MuiCollapse-hidden"
      style="min-height: 0px;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            role="region"
          >
            <div
              class="MuiAccordionDetails-root"
              style="margin: auto; padding: 8px 0px;"
            >
              <div
                class="MuiBox-root MuiBox-root-18"
                style="gap: 8px;"
              >
                <div
                  class="MuiBox-root MuiBox-root-19 ohm-pairs"
                  style="width: 100%;"
                >
                  <button
                    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-disableElevation MuiButton-fullWidth"
                    tabindex="0"
                    type="button"
                  >
                    <span
                      class="MuiButton-label"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1"
                      >
                        Add to Wallet
                      </p>
                    </span>
                  </button>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
